src/ssh/nativeSSH.ts

```diff
@@ -0,0 +1,111 @@
+import * as crypto from 'crypto';
+
+export class NativeSSHError extends Error {
+    constructor(code: number, stdout: string, stderr: string) {
+        super();
+        this.name = 'NativeSSHError';
+        this.message = `code: ${code}\n\nstdout: ${stdout}\n\nstderr: ${stderr}`;
+    }
+}
+
+function getSSHConfigPath() {
+    const sshPath = vscode.workspace.getConfiguration('remote.SSH').get<string>('path');
+    return sshPath || 'ssh';
+}
+
+function execCommand(commmad: string, args?: string[], options?: { timeout?: number }) {
+    const process = cp.spawn(commmad, args, { ...options, windowsVerbatimArguments: true });
+    const stdoutDataArr: string[] = [];
+    const stderrDataArr: string[] = [];
+    process.stdout.on('data', (data) => {
+        stdoutDataArr.push(data.toString());
+    });
+    process.stderr.on('data', (data) => {
+        stderrDataArr.push(data.toString());
+    });
+    const completed = new Promise<{ code: number }>((resolve, reject) => {
+        process.on('error', (err) => {
+            reject(err);
+        });
+        process.on('close', (code) => {
+            resolve({ code: code ?? 0 });
+        });
+    })
+
+    return {
+        get stdout() {
+            return stdoutDataArr.join();
+        },
+        get stderr() {
+            return stderrDataArr.join();
+        },
+        completed,
+        terminate() { process.kill(); }
+    }
+}
+
+let version: string | undefined;
+export async function getOpenSSHVersion(): Promise<string | undefined> {
+    if (version) {
+        return version;
+    }
+
+    try {
+        const sshPath = getSSHConfigPath();
+        const resp = execCommand(sshPath, ['-V'], { timeout: 3000 });
+        const { code } = await resp.completed;
+        if (code === 0) {
+            const match = /\bOpenSSH[A-Za-z0-9_\-\.]+\b/.exec(resp.stderr.trim() || resp.stderr.trim());
+            if (match) {
+                version = match[0];
+                return version;
+            }
+        }
+    } catch {
+    }
+    return undefined;
+}
+
+export async function getHostConfig(hostname: string) {
+    try {
+        const sshPath = getSSHConfigPath();
+        const resp = execCommand(sshPath, ['-T', '-G', hostname], { timeout: 3000 });
+        const { code } = await resp.completed;
+        if (code === 0) {
+            return resp.stdout.trim().split(/\r?\n/).map(s => s.trim());
+        }
+    } catch {
+    }
+    return undefined;
+}
+
+/**
+ * Test ssh connection to a remote machine using native ssh client.
+ *
+ * Caller should handle any exception thrown
+ */
+export async function testSSHConnection(username: string, hostname: string) {
+    const sshPath = getSSHConfigPath();
+    const randomId = crypto.randomBytes(12).toString('hex');
+    // TODO: support password input so we can replace the ssh2 library
+    const resp = execCommand(sshPath, ['-T', '-o', 'ConnectTimeout=5', `${username}@${hostname}`, `echo "${randomId}"`], { timeout: 5500 });
```

Comment:
Will `timeout 5s` too short? since we have `10s` of retry in proxy itself, in my network condition, it may cost 3+ seconds (depend on provider state)

Reply:
We are afraid too block unnecessary users on SSH Gateway.

The plan is to rollback windows users to SSH Gateway and then do dry runs. I would suggest to go with small timeout and if we see a lot of timeout issues then we increase it.